ui/vty/src/Ariadne/UI/Vty/Widget.hs

```diff
@@ -0,0 +1,464 @@
+-- Types and instances
+----------------------------------------------------------------------------
+
+-- | Component of a unique widget name
+--
+-- Each widget has its child widgets indexed by this name component.
+-- Widget's unique name is constructed as a list of these components,
+-- corresponding to its location in widget tree.
+--
+-- Should be made into a union of widget-local name types.
+data WidgetNamePart
+  = WidgetNameSelf  -- Special name for adding widget itself to focus list
+
+  | WidgetNameMenu
+  | WidgetNameTree
+  | WidgetNameStatus
+  | WidgetNameHelp
+  | WidgetNameAbout
+  | WidgetNameLogs
+
+  | WidgetNameRepl
+  | WidgetNameReplInput
+
+  | WidgetNameAddWallet
+  | WidgetNameAddWalletNewName
+  | WidgetNameAddWalletNewPass
+  | WidgetNameAddWalletNewButton
+  | WidgetNameAddWalletRestoreName
+  | WidgetNameAddWalletRestoreMnemonic
+  | WidgetNameAddWalletRestorePass
+  | WidgetNameAddWalletRestoreFull
+  | WidgetNameAddWalletRestoreButton
+
+  | WidgetNameWallet
+  | WidgetNameWalletSendAddress
+  | WidgetNameWalletSendAmount
+  | WidgetNameWalletSendPass
+  | WidgetNameWalletSendButton
+
+  | WidgetNameAccount
+  deriving (Eq, Ord, Show)
+
+-- | Unique widget name, describing its location in widget tree
+type WidgetName = [WidgetNamePart]
+
+-- | Events child widget can send back to parent widget
+--
+-- Should also be a union type of widget-local event types
+data WidgetEvent
+  = WidgetEventMenuSelected
+  | WidgetEventButtonPressed
+  | WidgetEventEditChanged
+
+-- | Result of handling a UI event
+--
+-- Each UI event is sent to the deepest focused widget possible.
+-- If its handler returns @WidgetEventNotHandled@, the event "bubbles" up
+-- through the widget hierarchy and is fed to parent widget's event handler.
+data WidgetEventResult
+  = WidgetEventHandled
+  | WidgetEventNotHandled
+
+-- | Base widget type
+--
+-- Contains local widget state @s@, child widgets, drawing routines and event handlers.
+-- It is parametrized by parent's @WidgetInfo@ type to allow access to its state
+-- through lenses, and also its event queue for child→parent communication via @WidgetEvent@s.
+data WidgetInfo s p = WidgetInfo
+  { widgetName :: !WidgetName
+  , widgetState :: s
+  -- ^ Lazy, because can be undefined for stateless widgets
+  , widgetChildren :: !(Map WidgetNamePart (Widget (WidgetInfo s p)))
+  , widgetFocusList :: ![WidgetNamePart]
+  , widgetEventHandlers :: !(Map WidgetNamePart (WidgetEvent -> WidgetEventM s p ()))
+  -- ^ Should be refactored into a dependent-map, for something like this:
+  -- @Map (e, WidgetNamePart) -> (e -> m ())@
+  , widgetEventQueue :: ![(WidgetNamePart, WidgetEvent)]
+  -- ^ List of events received from child widgets, is traversed after child event handlers complete
+  , widgetEventSend :: !(WidgetEvent -> WidgetEventM s p ())
+  -- ^ Send event to parent widget's queue
+  , widgetDraw :: !(s -> WidgetDrawM s p (B.Widget WidgetName))
```

Comment:
(nitpick)
Why do you need to pass `s` as argument when `WidgetDrawM` is a reader over this struct, so one can define

```haskell
getState :: WidgetDrawM s p s
getState = view widgetStateL
```